src/script.js

```diff
@@ -251,54 +331,90 @@ function isMultisigInput (script, allowIncomplete) {
+  if (chunks.length < 4) {
+    throw new Error('Multisig script should contain at least 4 elements')
+  }
+  if (chunks[chunks.length - 1] !== OPS.OP_CHECKMULTISIG) {
+    throw new Error('Final opcode should be OP_CHECKMULTISIG')
+  }
+  if (!types.Number(chunks[0])) {
+    throw new Error('First element must be a number-push opcode')
+  }
+  if (!types.Number(chunks[chunks.length - 2])) {
+    throw new Error('Second-last element must be a number-push opcode')
+  }
+  var m = chunks[0] - OP_INT_BASE
+  var n = chunks[chunks.length - 2] - OP_INT_BASE
 
-  if (!types.Number(mOp)) return false
-  if (!types.Number(nOp)) return false
+  if (m <= 0) {
+    throw new Error('Number of signatures required must be greater than zero')
+  }
+  if (n > 16) {
+    throw new Error('Number of public keys must be less than 17')
+  }
+  if (m > n) {
+    throw new Error('Number of signatures cannot exceed the number of public keys')
+  }
+  if (n !== chunks.length - 3) {
+    throw new Error('n does not match the number of public keys')
+  }
 
-  var m = mOp - OP_INT_BASE
-  var n = nOp - OP_INT_BASE
+  var keys = chunks.slice(1, -2)
+  if (!keys.every(isCanonicalPubKey)) {
+    throw new Error('Non-canonical pubic key found')
+  }
 
-  // 0 < m <= n <= 16
-  if (m <= 0) return false
-  if (m > n) return false
-  if (n > 16) return false
-  if (n !== chunks.length - 3) return false
+  return {
+    type: 'multisig',
+    nRequiredSigs: m,
+    nPublicKeys: n,
+    publicKeyBuffers: keys
+  }
+}
 
-  return chunks.slice(1, -2).every(isCanonicalPubKey)
+function isMultisigOutput (script) {
+  return determinesTypeOrNonstandard([decodeMultisigOutput], script) === 'multisig'
 }
 
 function isNullDataOutput (script) {
-  var chunks = decompile(script)
-  return chunks[0] === OPS.OP_RETURN
+  return determinesTypeOrNonstandard([decodeNullDataOutput], script) === 'nulldata'
 }
 
-function classifyOutput (script) {
-  var chunks = decompile(script)
+function decodeNullDataOutput (script, chunks) {
+  if (script[0] !== OPS.OP_RETURN) {
+    throw new Error('Missing OP_RETURN at start of script')
+  }
 
-  if (isWitnessPubKeyHashOutput(chunks)) {
-    return 'witnesspubkeyhash'
-  } else if (isWitnessScriptHashOutput(chunks)) {
-    return 'witnessscripthash'
-  } else if (isPubKeyHashOutput(chunks)) {
-    return 'pubkeyhash'
-  } else if (isScriptHashOutput(chunks)) {
-    return 'scripthash'
-  } else if (isMultisigOutput(chunks)) {
-    return 'multisig'
-  } else if (isPubKeyOutput(chunks)) {
-    return 'pubkey'
-  } else if (isNullDataOutput(chunks)) {
-    return 'nulldata'
+  return {
+    type: 'nulldata'
   }
+}
 
-  return 'nonstandard'
+function determinesTypeOrNonstandard (functions, script) {
+  if (!types.Array(functions)) {
+    throw new Error('Must provide an array of functions to determinesTypeOrNonstandard')
+  }
+  if (!types.Buffer(script)) {
+    throw new Error('Must provide a script to determinesTypeOrNonstandard')
```

Comment:
Right, so this is why this is a breaking change.

Why can't we just coerce from `Buffer`?
Also,  use `typeforce`... :+1: